src/gsym/parser.rs

```diff
@@ -0,0 +1,330 @@
+        off += 8;
+        let num_addrs = decode_uword(&data[off..]);
+        off += 4;
+        let strtab_offset = decode_uword(&data[off..]);
+        off += 4;
+        let strtab_size = decode_uword(&data[off..]);
+        off += 4;
+        let uuid: [u8; 20] = (&data[off..(off + 20)])
+            .try_into()
+            .expect("input data is too short");
+        off += 20;
+
+        // Get the slices of the Address Table, Address Data Offset Table,
+        // and String table.
+        let addr_tab = &data[off..(off + num_addrs as usize * addr_off_size as usize)];
+        off += num_addrs as usize * addr_off_size as usize;
+        let addr_data_off_tab = &data[off..(off + num_addrs as usize * 4)];
+        off += num_addrs as usize * 4;
+        let str_tab =
+            &data[strtab_offset as usize..(strtab_offset as usize + strtab_size as usize)];
+
+        Ok(GsymContext {
+            header: Header {
+                magic,
+                version,
+                addr_off_size,
+                uuid_size,
+                base_address,
+                num_addrs,
+                strtab_offset,
+                strtab_size,
+                uuid,
+            },
+            addr_tab,
+            addr_data_off_tab,
+            str_tab,
+            raw_data: data,
+        })
+    }
+
+    #[inline(always)]
+    pub fn num_addresses(&self) -> usize {
+        self.header.num_addrs as usize
+    }
+
+    /// Get the address of the an entry in the Address Table.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid index.
+    #[inline(always)]
+    pub fn addr_at(&self, idx: usize) -> u64 {
+        assert!(idx < self.header.num_addrs as usize, "invalid index");
+        let off = idx * self.header.addr_off_size as usize;
+        let mut addr = 0u64;
+        let mut shift = 0;
+        for d in &self.addr_tab[off..(off + self.header.addr_off_size as usize)] {
+            addr |= (*d as u64) << shift;
+            shift += 8;
+        }
+        addr += self.header.base_address;
+        addr
+    }
+
+    /// Get the AddressInfo of an address given by an index.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid index.
+    #[inline(always)]
+    pub fn addr_info(&self, idx: usize) -> AddressInfo {
+        assert!(idx < self.header.num_addrs as usize, "invalid index");
+        let off = idx * 4;
+        let ad_off = decode_uword(&self.addr_data_off_tab[off..]) as usize;
+        let size = decode_uword(&self.raw_data[ad_off..]);
+        let name = decode_uword(&self.raw_data[ad_off + 4..]);
+        AddressInfo {
+            size,
+            name,
+            data: &self.raw_data[ad_off + 8..],
+        }
+    }
+
+    /// Get the string at the given offset from the String Table.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid offset.
+    #[inline(always)]
+    pub fn get_str(&self, off: usize) -> &str {
```

Comment:
Could we return a `CStr` instead? Or at least not panic on malformed stuff by using `Option<&str>`?